kvserver: remove legacy snapshot and diff code

pav-kv opened the pull request:
This commit removes the proto and code for `RaftDataSnapshot`. This type was used for reporting replica snapshots and computing diffs, but now this functionality has been removed in favor of storage checkpoints and offline tooling.

Part of #21128
Epic: none

Release note: None

pav-kv committed:
kvserver: remove legacy snapshot and diff code

This commit removes the proto and code for RaftDataSnapshot. This type was used
for reporting replica snapshots and computing diffs, but now this functionality
has been removed in favor of storage checkpoints and offline tooling.

Release note: None

pavelkalinnikov commented on pkg/kv/kvserver/replica_consistency.go:
```diff
@@ -482,18 +480,18 @@ type replicaHash struct {
-	return snap, nil
+// ChecksumRangeForTesting returns a checksum over the KV data of the given
+// range. Only for testing.
+func ChecksumRangeForTesting(
```
Move this to `_test.go` file, as @tbg suggested previously.

erikgrinaker commented:
Let's hold off until we know what's up with 22.1 compatibility. I think we can still get rid of the diffs, because 22.1 will degrade gracefully if we omit them from the responses, but let's confirm.

erikgrinaker commented on pkg/kv/kvserver/helpers_test.go:
```diff
@@ -550,3 +550,16 @@ func WatchForDisappearingReplicas(t testing.TB, store *Store) {
+func ChecksumRange(
+	ctx context.Context, desc roachpb.RangeDescriptor, snap storage.Reader,
+) ([]byte, error) {
+	var r *Replica // TODO(pavelkalinnikov): make this less ugly.
```
`sha512()` doesn't use the `Replica` descriptor for anything, so we can just rename and export it instead.

pavelkalinnikov replied:
I'll see tomorrow if I can make this even cleaner, and will send a follow-up PR if so.

erikgrinaker reviewed (approved):
Good riddance!

As you may know, version-skipping upgrades were pushed back, so we don't need 22.1 compatibility for 23.1. In other words, let's merge this.

pav-kv committed:
kvserver: make replica sha512 a function

This function does not use the replica, so does not have to be a method.

Epic: None
Release note: None

pavelkalinnikov commented:
bors r=erikgrinaker
